imap/index.c

```diff
@@ -5538,50 +5429,50 @@ static int extract_attachment(const char *type, const char *subtype,
         if (r) {
             syslog(LOG_ERR,
                    "extract_attachment: failed to read response for GET %s/%s",
-                   ctx->path, guidstr);
-            r = IMAP_IOERROR;
-            goto done;
+                   ext->path, guidstr);
+            statuscode = 599;
         }
     } while (statuscode < 200);
 
     syslog(LOG_DEBUG, "extract_attachment: GET %s/%s: got status %u",
-           ctx->path, guidstr, statuscode);
+           ext->path, guidstr, statuscode);
 
-    /* Abide by server's timeout, if any */
-    if (!ctx->timeout) {
-        const char **hdr, *p;
+    if (statuscode == 200) goto gotdata;
 
-        if ((hdr = spool_getheader(hdrs, "Keep-Alive")) &&
-            (p = strstr(hdr[0], "timeout="))) {
-            ctx->timeout = atoi(p+8);
-        }
-        else ctx->timeout = IDLE_TIMEOUT;
+    // otherwise we're going to try three times to PUT this request to the server!
 
-        if (be->timeout) be->timeout->mark = time(NULL) + ctx->timeout;
+    /* Decode data */
+    struct buf decbuf = BUF_INITIALIZER;
+    if (encoding) {
+        if (charset_decode(&decbuf, buf_base(data), buf_len(data), encoding)) {
+            syslog(LOG_ERR, "extract_attachment: failed to decode data");
+            r = IMAP_IOERROR;
+            goto done;
+        }
+        data = &decbuf;
     }
 
-    if (statuscode == 404) {
-        /* Decode data */
-        struct buf decbuf = BUF_INITIALIZER;
-        if (encoding) {
-            if (charset_decode(&decbuf, buf_base(data), buf_len(data), encoding)) {
-                syslog(LOG_ERR, "extract_attachment: failed to decode data");
-                r = IMAP_IOERROR;
-                goto done;
-            }
-            data = &decbuf;
+    /* Build list of Content-Type parameters */
+    const struct param *param = type_params;
+    while (param && param->attribute) {
+        buf_putc(&buf, ';');
+        buf_appendcstr(&buf, param->attribute);
+        if (param->value) {
+            buf_putc(&buf, '=');
+            buf_appendcstr(&buf, param->value);
         }
+        param = param->next;
+    }
 
-        /* Build list of Content-Type parameters */
-        const struct param *param = type_params;
-        while (param && param->attribute) {
-            buf_putc(&buf, ';');
-            buf_appendcstr(&buf, param->attribute);
-            if (param->value) {
-                buf_putc(&buf, '=');
-                buf_appendcstr(&buf, param->value);
-            }
-            param = param->next;
+    int retry;
+    for (retry = 0; retry < 3; retry++) {
+        if (retry) {
+            // second and third time around, sleep and reconnect
+            sleep(retry);
```

Comment:
This means 3 seconds total sleeping, 1 then 2.  We disconnect each time, which means that the following reconnect will be a real reconnect!

Reply:
We could also not bail out of this loop early if extractor_connect failed, e.g. if the network was down for a second.